server/tserver/src/main/java/org/apache/accumulo/tserver/ThriftClientHandler.java

```diff
@@ -0,0 +1,1812 @@
+      throws ThriftSecurityException {
+    try {
+      log.trace("Got {} message from user: {}", request, credentials.getPrincipal());
+      if (!security.canPerformSystemActions(credentials)) {
+        log.warn("Got {} message from user: {}", request, credentials.getPrincipal());
+        throw new ThriftSecurityException(credentials.getPrincipal(),
+            SecurityErrorCode.PERMISSION_DENIED);
+      }
+    } catch (ThriftSecurityException e) {
+      log.warn("Got {} message from unauthenticatable user: {}", request, e.getUser());
+      if (server.getContext().getCredentials().getToken().getClass().getName()
+          .equals(credentials.getTokenClassName())) {
+        log.error("Got message from a service with a mismatched configuration."
+            + " Please ensure a compatible configuration.", e);
+      }
+      throw e;
+    }
+
+    if (server.getLock() == null || !server.getLock().wasLockAcquired()) {
+      log.debug("Got {} message before my lock was acquired, ignoring...", request);
+      throw new RuntimeException("Lock not acquired");
+    }
+
+    if (server.getLock() != null && server.getLock().wasLockAcquired()
+        && !server.getLock().isLocked()) {
+      Halt.halt(1, () -> {
+        log.info("Tablet server no longer holds lock during checkPermission() : {}, exiting",
+            request);
+        server.getGcLogger().logGCInfo(server.getConfiguration());
+      });
+    }
+
+    if (lock != null) {
+      ZooUtil.LockID lid =
+          new ZooUtil.LockID(server.getContext().getZooKeeperRoot() + Constants.ZMASTER_LOCK, lock);
+
+      try {
+        if (!ZooLock.isLockHeld(server.getMasterLockCache(), lid)) {
+          // maybe the cache is out of date and a new master holds the
+          // lock?
+          server.getMasterLockCache().clear();
+          if (!ZooLock.isLockHeld(server.getMasterLockCache(), lid)) {
+            log.warn("Got {} message from a master that does not hold the current lock {}", request,
+                lock);
+            throw new RuntimeException("bad master lock");
+          }
+        }
+      } catch (Exception e) {
+        throw new RuntimeException("bad master lock", e);
+      }
+    }
+  }
+
+  @Override
+  public void loadTablet(TInfo tinfo, TCredentials credentials, String lock,
+      final TKeyExtent textent) {
+
+    try {
+      checkPermission(credentials, lock, "loadTablet");
+    } catch (ThriftSecurityException e) {
+      log.error("Caller doesn't have permission to load a tablet", e);
+      throw new RuntimeException(e);
+    }
+
+    final KeyExtent extent = new KeyExtent(textent);
+    SortedSet<KeyExtent> unopenedTablets = server.getUnopenedTablets();
+    SortedSet<KeyExtent> openingTablets = server.getOpeningTablets();
+
+    synchronized (unopenedTablets) {
+      synchronized (openingTablets) {
+        SortedMap<KeyExtent,Tablet> onlineTablets = server.getOnlineTablets();
+        synchronized (onlineTablets) {
```

Comment:
This is changing the synchronization. Previously, it synchronized on the snapshot factory. Now, it's synchronizing on the snapshot. This kind of error is why I think it'd probably be better to use package-private references to TabletServer's own fields, rather than use getters and assign to new variables.

I saw you avoided this behavior change elsewhere by synchronizing on what is returned from your new method, `getOnlineTabletsRaw()`, but I think it's too easy to mistake these methods for one another.

Another reason to use the same reference is because it makes it easier to analyze the concurrency model by finding references the object in an IDE. If the same object is assigned to multiple references, trying to understand the concurrency model is more difficult, because you have to follow call stacks.

Reply:
Whoops, yea, the syncrhonziation isn't actually necessary since it's operating on the snapshot. Yea, as I mentioned in slack, the use of syncrhonzied colletions makes this a bit confusing and would be more straight forward to refactor them as concurrent collections later.

Reply:
Perhaps... making them concurrent instead of having the explicit synchronization... might make it harder to understand the overall concurrency model, and the relationship between these objects. It might make more sense to put the objects into a single container that gives a consistent snapshot across all of the data structures at once. I'd have to think about it some more to have an informed opinion.